Experiments with subnetworks

lucasosouza opened the pull request:
* notebook with experiments on creating subnetworks with dendrites (RES-1912), and utils.py with supporting code
* a notebook tutorial for Vernon (RES-1931)

ps: utils.py, which is under a projects folder, uses the nomenclature adopted by scikit-learn of X and Y for data and target respectively. Hence the noqa on that file. 

ps2: since it is also just a projects file, I am not adding scikit-learn to nupic.research requirements.

karangrewal commented on projects/dendrites/subnetworks/utils.py:
```diff
@@ -0,0 +1,383 @@
+
+
+def add_task_embedding(X, task_num=1, units=2):
+    """Adds a task embedding to the input"""
```
Can you add more detail in the docstring about what this function does?

lucasosouza replied:
Done

karangrewal commented on projects/dendrites/subnetworks/utils.py:
```diff
@@ -0,0 +1,383 @@
+
+
+def load_datasets():
+    """Load Iris, Wine and Mixed datasets"""
```
Can you please describe exactly what is returned by this function so that others will know how to call it appropriately?

lucasosouza replied:
Done

karangrewal commented on projects/dendrites/subnetworks/utils.py:
```diff
@@ -0,0 +1,383 @@
+    print(f"{dataset_name} Acc: {accuracy_score(target, pred):.4f}")
+
+def train(model, data, target, dataset_name, batch_size=1,
+          epochs=10, lr=1e-2, weight_decay=0, verbose=True):
```
how is this training function different from the one in [model utils](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/model_utils.py#L37-L183), and why can't we re-use that?

lucasosouza replied:
Not really different, but this is a small self-contained experiment. I needed the flexibility to modify the function as needed.

karangrewal commented on projects/dendrites/subnetworks/utils.py:
```diff
@@ -0,0 +1,383 @@
+# flake8: noqa
```
Please include the software license at the top of this file.

lucasosouza replied:
Good catch

karangrewal reviewed (changes requested):
I requested some more documentation in a few areas, just to help future users who want to call your util functions.

Also, you defined `StandardMLP` and `DendriticMLP` which copy a lot of code, and the only difference between these and the networks defined in frameworks is additional batch normalization layers (as far as I could tell). I'm not sure if it's better to redefine these architectures here or in frameworks, so I will let you decide.

lucasosouza commented:
Thanks @karangrewal , really good points. I incorporated your comments and replied above. I also fixed all styles issues in the document except for the variable names, as mentioned.
